calico/felix/config.py

```diff
@@ -46,166 +46,299 @@
+
+        if final:
+            # Update logging.
+            common.complete_logging(self.LOGFILE,
+                                    self.LOGLEVFILE,
+                                    self.LOGLEVSYS,
+                                    self.LOGLEVSCR)
+
+            # Log configuration - the whole lot of it.
+            for name, parameter in self.parameters.iteritems():
+                log.info("Parameter %s (%s) has value %r read from %s",
+                         name,
+                         parameter.description,
+                         parameter.value,
+                         parameter.active_source)
+
+    def _read_env_vars(self):
+        """
+        Read all of the variables from the environment.
+        """
+        for name, parameter in self.parameters.iteritems():
+            # All currently defined config parameters have ENV as a valid source.
+            assert(ENV in parameter.sources)
+            # ENV is the first source, so we can assert that using defaults.
+            assert(parameter.active_source is None)
+
+            env_var = ("FELIX_%s" % name).upper()
+            if env_var in os.environ:
+                parameter.set(os.environ[env_var],
+                              "Environment variable %s" % env_var)
+
+    def _read_cfg_file(self, config_file):
+        parser = ConfigParser.ConfigParser()
+        parser.read(config_file)
+        cfg_dict = {}
+
+        # Build up the cfg dictionary from the file.
+        for section in parser.sections():
+            cfg_dict.update(dict(parser.items(section)))
+
+        source = "Configuration file %s" % config_file
+
+        for name, parameter in self.parameters.iteritems():
+            # Config parameters are lower-cased by ConfigParser
+            name = name.lower()
+            if FILE in parameter.sources and name in cfg_dict:
+                # This can validly be read from file.
+                parameter.set(cfg_dict.pop(name), source)
+        self._warn_unused_cfg(cfg_dict, source)
+
+    def report_etcd_config(self, host_dict, global_dict):
+        """
+        Report back configuration parameters read from etcd.
+        :param host_dict: Dictionary of etcd parameters
+        :param global_dict: Dictionary of global parameters
+        :raises ConfigException
+        """
+        log.debug("Configuration reported from etcd")
+        for source in (LOCAL_ETCD, GLOBAL_ETCD):
+            if source == LOCAL_ETCD:
+                cfg_dict = host_dict
+            else:
+                cfg_dict = global_dict
+
+            for name, parameter in self.parameters.iteritems():
+                if source in parameter.sources and name in cfg_dict:
+                    parameter.set(cfg_dict.pop(name), source)
 
-        self._config_path = config_path
-        self._items = {}
+            self._warn_unused_cfg(cfg_dict, source)
 
-        self.read_cfg_file(config_path)
+        self._finish_update(final=True)
```

Comment:
This function has a bad name and a bad docstring. 

- "Report back configuration parameters read from etcd": report them back to where?
- Can I run this function multiple times? It calls something called `_finish_update`: what does that have to do with anything?